src/request.ts

```diff
@@ -0,0 +1,362 @@
+/**
+ * Type safe wrapper around the fetch API
+ */
+
+// TODO: add body type parameter to POST/PUT
+// TODO: support for optional query parameters
+// TODO: when query/headers type is "never", it should not allow any query/header to be produced
+// TODO: add timeout to fetch (how to cancel request?)
```

Comment:
consider also using something like superagent instead of fetch which supports timeouts and retry

Reply:
it's not clear, how's p canceled ? (and the fetch operation ?)

Reply:
I'm not sure you have control over that, if we really want the request to stop mid air because we say so, I guess we can use [this approach](https://developers.google.com/web/updates/2017/09/abortable-fetch) but I need to check if `fetch` API supports that.

Anyway, the `createFetchRequestForApi` can take any function that implements the `fetch` API, so you can implement something more sophisticated using another library or lower level APIs.